### ISSUE TITLE
git.latest and git.present state functions are misleading

### ISSUE THREAD

**Issue by ryan-lane** [saltstack/salt#13550]
git.latest will do a clone and always will fetch is no rev is listed, but will only clone and never update if rev is given and always_fetch is False (the default). git.present will just create an empty repo. At minimum git.latest is poorly named.

**Comment by anlutro** [comment]
I'm having a problem with `git.latest` that I think would benefit from a `git.cloned` state. I set up bare repositories on my server that I can push to, then I clone the bare repository elsewhere on the server and set up a post-update hook in the bare repository to pull changes in.

The following state SLS will fail because salt tries doing `git rev-parse HEAD` in the cloned repository, which fails.

```yaml
/tmp/git_bare:
  git.present:
    - bare: true

/tmp/git_cloned:
  git.latest:
    - name: /tmp/git_bare
    - target: /tmp/git_cloned

# omitted: setting up the post-update hook
```

**Comment by jfindlay** [comment]
Since the pull request has been merged, I'm going to consider this issue closed.

**Comment by anlutro** [comment]
@jfindlay The pull request that mentions this issue isn't really a fix to this issue. I just mentioned it as vaguely related in a comment, sorry for the confusion.

I still think a `git.cloned` state would be useful as an alternative to `git.latest`.

**Comment by terminalmage** [comment]
@ryan-lane There is an open issue for expanding the arguments in git.present to more closely match git.latest (https://github.com/saltstack/salt/issues/13317). The idea for this would be to make git.present work more like pkg.installed, while git.latest would work more like pkg.latest. That is, git.present would clone the repo and rev if there is no git repo at the target location, but it would not keep the git checkout up-to-date.

The work on git.present is slated for the "Boron" feature release.

As for git.latest, I've rewritten it completely for the 2015.8.0 release (pull request [here](https://github.com/saltstack/salt/pull/26341)). I almost decided to change the default value of ``always_fetch`` from ``False`` to ``True``, but after reading this issue I decided to deprecate this parameter altogether (see https://github.com/saltstack/salt/pull/26603). I totally agree that this state implies that it keeps the local checkout up-to-date, and it wasn't always doing that. Also, part of the reason that ``always_fetch`` existed in the first place is because the logic for when the state does a fetch wasn't all that robust, and my rewrite has addressed this pretty well. It makes use of ``git ls-remote`` now to get the SHA1 of the remote revision, and checks to see if the revision exists locally. If it does, no fetch is performed.

Please let me know if my recent rewrite of git.latest, and the proposed changes to git.present, together address your concerns. If so, we can close this issue, and continue to track the progress of the upcoming git.present changes in https://github.com/saltstack/salt/issues/13317.

**Comment by akoumjian** [comment]
For the current release (2015.5.5) what is the recommended state to put a git clone into the latest version of a branch? I have this and it is no longer working since I upgraded:

```
webapp:
  git.latest:
    - name: git@github.com:org/repo.git
    - rev: {{ git_rev }}
    - target: /var/www/
    - force: True
    - force_checkout: True
    - force_reset: True
    - always_fetch: True
    - require:
      - pkg: git
      - pkg: app-pkgs
      - file: /var/www
      - file: ssh_config
      - file: deploykey
      - file: publickey
```

**Comment by SunPowered** [comment]
I'm just getting to know saltstack, and as a new user my impression of the commands were as follows:

`git.latest` - Always keeps up to date with a pull request to the repository, unless a rev is explicitly named
`git.present` - Should take the same params as `git.latest`, but will clone and checkout only if the repo does not already exist locally.

Reading into this  in more detail, I found the actual behaviour somewhat non-intuitive.

my 2c

**Comment by kojiromike** [comment]
I think if you use `git.latest` and `rev` is `HEAD` (the default), then `git.latest` should attempt to update the worktree every time that state applies. (To be sure, even if `rev` is explicitly defined to be `HEAD`, I think a pull is still in order.)

**Comment by terminalmage** [comment]
@akoumjian Sorry I missed your reply back when you made it. Your SLS should work, though the ``force`` param is no longer used in 2015.8.0 and later. If you're saying that your SLS doesn't work, then I'd need more information to be able to look into it.

@kojiromike It does. That's the entire point of the state. It doesn't do a pull though, it fetches first and does some checks to discover if the update is a fast-forward. If it is a fast-forward, then a ``git merge`` is done to fast-forward the repo, and if not then either A) the state exits with a ``False`` result if ``force_reset`` is set to ``False``, or does a ``git reset --hard`` to the SHA of the specified ref if ``force_reset`` is set to ``True``.

**Comment by akoumjian** [comment]
That is correct, it does not work. @terminalmage what information do you need?

Steps to reproduce:

1. Specify a git.latest state using the rev argument to point to a branch with any possible combination of the force_* options
2. add some local changes to the git directory, that should be discarded on the next state.highstate
3.  run state.highstate, see that it does not discard local changes and pull latest

**Comment by terminalmage** [comment]
@akoumjian What version are you running? This seems similar to https://github.com/saltstack/salt/issues/34725, which you opened a few months ago, and which I submitted a fix for that is in 2016.3.2 and later. I also included a regression test for the fix.

**Comment by akoumjian** [comment]
This will take some time for me to re-set up and test on latest. I've been using a fork of the git state for a while now. It's fully possible it's been fixed on latest for a while and I haven't seen it. Forgot about the other ticket.
